src/config/groups/keys.cpp

```diff
@@ -0,0 +1,1594 @@
+
+    if (0 != crypto_scalarmult_ed25519_noclamp(kT.data(), k.data(), T.data()))
+        throw std::runtime_error{"scalarmult failed: perhaps an invalid session id?"};
+
+    ustring out;
+    out.resize(4 + 32 + 64);
+    out[0] = 0x03;                      // network prefix
+    out[1] = subacc_flags(write, del);  // permission flags
+    out[2] = 0;                         // reserved 1
+    out[3] = 0;                         // reserved 2
+    // The next 32 bytes are k (NOT kT; the user can go make kT themselves):
+    std::memcpy(&out[4], k.data(), k.size());
+
+    // And then finally, we append a group signature of: p || f || 0 || 0 || kT
+    std::array<unsigned char, 36> to_sign;
+    std::memcpy(&to_sign[0], out.data(), 4);  // first 4 bytes are the same as out
+    std::memcpy(&to_sign[4], kT.data(), 32);  // but then we have kT instead of k
+    crypto_sign_ed25519_detached(&out[36], nullptr, to_sign.data(), to_sign.size(), c.data());
+
+    return out;
+}
+
+ustring Keys::swarm_subaccount_token(std::string_view session_id, bool write, bool del) const {
+    if (!admin())
+        throw std::logic_error{"Cannot make subaccount signature: admin keys required"};
+
+    // Similar to the above, but we only care about getting flags || kT
+
+    auto X = session_id_pk(session_id);
+    auto& c = _sign_sk;
+    auto& C = *_sign_pk;
+
+    auto k = subaccount_blind_factor(X);
+
+    // T = |S|
+    auto T = xed25519::pubkey(ustring_view{X.data(), X.size()});
+
+    ustring out;
+    out.resize(4 + 32);
+    out[0] = 0x03;                      // network prefix
+    out[1] = subacc_flags(write, del);  // permission flags
+    out[2] = 0;                         // reserved 1
+    out[3] = 0;                         // reserved 2
+    if (0 != crypto_scalarmult_ed25519_noclamp(&out[4], k.data(), T.data()))
+        throw std::runtime_error{"scalarmult failed: perhaps an invalid session id?"};
+    return out;
+}
+
+Keys::swarm_auth Keys::swarm_subaccount_sign(
+        ustring_view msg, ustring_view sign_val, bool binary) const {
+    if (sign_val.size() != 100)
+        throw std::logic_error{"Invalid signing value: size is wrong"};
+
+    if (!_sign_pk)
+        throw std::logic_error{"Unable to verify: group pubkey is not set (!?)"};
+
+    Keys::swarm_auth result;
+    auto& [token, sub_sig, sig] = result;
+
+    // (see above for variable/crypto notation)
+
+    ustring_view k = sign_val.substr(4, 32);
+
+    // our token is the first 4 bytes of `sign_val` (flags, etc.), followed by kT which we have to
+    // compute:
+    token.resize(36);
+    std::memcpy(token.data(), sign_val.data(), 4);
+
+    // T = |S|, i.e. we have to clear the sign bit from our pubkey
+    std::array<unsigned char, 32> T;
+    crypto_sign_ed25519_sk_to_pk(T.data(), user_ed25519_sk.data());
+    bool neg = T[31] & 0x80;
+    T[31] &= 0x7f;
+    if (0 != crypto_scalarmult_ed25519_noclamp(to_unsigned(token.data() + 4), k.data(), T.data()))
```

Comment:
The short answer is that it "doesn't work" with clamping half the time, because when we have to use `-a` instead of `a`, clamping it changes the value and breaks the `-A = -aG` relationship that we're depending on.  There's a fair bit to unpack behind this though, so let me elaborate.

I wish we didn't need that relationship at all: but unfortunately it's an artifact of a suboptimal choice (which unfortunately followed Signal) of having Session IDs be the X25519 pubkey conversions of Ed25519, rather than direct Ed25519 keys.  Unfortunately that Ed->X conversion loses the sign bit from the Ed25519 pubkey, which is what causes all this trouble: when reversing the Ed->X conversion we get ourselves to Ed=±sqrt(...) and there's no way to know which one was the correct Ed pubkey.  So what we're doing here is effectively always signing *as if* the real pubkey that yielded the Session ID was in fact always the positive of the two possible Ed25519 pubkeys, when in reality half the time it is negative.

In order to do that when we actually have a negative pubkey, however, we have to effectively produce a signature from `-a` rather than from `a` -- but that won't work if we clamp `-a`.

To illustrate:

```Python
#!/usr/bin/python3

from nacl.signing import SigningKey
import nacl.bindings as sodium

for seed31 in (b'\0', b'\4'):
    seed = SigningKey(b"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0" + seed31)
    pubkey = seed.verify_key

    # This gives the private scalar, i.e. "a" for keypair a/A that yields aG = A.  (That's not its primary usage in
    # libsodium, and thus the function name here is understandably confusing).
    a = seed.to_curve25519_private_key().encode()

    print(f"\nPubkey:     {pubkey.encode().hex()}")
    Ac = sodium.crypto_scalarmult_ed25519_base(a)
    An = sodium.crypto_scalarmult_ed25519_base_noclamp(a)
    print(f"A noclamp:  {An.hex()}")
    print(f"A clamped:  {Ac.hex()}")

    if Ac[31] & 0x80:
        a_neg = sodium.crypto_core_ed25519_scalar_negate(a)
        Ac_neg = sodium.crypto_scalarmult_ed25519_base(a_neg)
        An_neg = sodium.crypto_scalarmult_ed25519_base_noclamp(a_neg)
        print(f"-A noclamp: {An_neg.hex()}")
        print(f"-A clamped: {Ac_neg.hex()}")
```
produces:
```
Pubkey:     3b6a27bcceb6a42d62a3a8d02a6f0d73653215771de243a63ac048a18b59da29
A noclamp:  3b6a27bcceb6a42d62a3a8d02a6f0d73653215771de243a63ac048a18b59da29
A clamped:  3b6a27bcceb6a42d62a3a8d02a6f0d73653215771de243a63ac048a18b59da29

Pubkey:     fd50b8e3b144ea244fbf7737f550bc8dd0c2650bbc1aada833ca17ff8dbf329b
A noclamp:  fd50b8e3b144ea244fbf7737f550bc8dd0c2650bbc1aada833ca17ff8dbf329b
A clamped:  fd50b8e3b144ea244fbf7737f550bc8dd0c2650bbc1aada833ca17ff8dbf329b
-A noclamp: fd50b8e3b144ea244fbf7737f550bc8dd0c2650bbc1aada833ca17ff8dbf321b
-A clamped: d1cd8406d8f93617b10a91a20c4126c08bb4f0fb23807fcfd25e939b7a9924ea
```
(Keep in mind these are encoded in little-endian, and so the sign bit is the most significant bit of the *last* byte, i.e. the second-last hex digit).

So in short, without clamping we can keep up the illusion that we were actually the positive alternative of the two pubkeys that could have given the X25519, even when we aren't.

Reply:
> Do you think we should clamp just in case we decide to reuse this blinded keypair for X25519 key exchange in the future for some reason ?

I really wish we could, because you're right, we're okay for Ed25519 signatures, but definitely don't want to go back to X again with this derived key.